cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -308,11 +309,24 @@ class DatasetWriterDirectoryQueue {
   }
 
   Result<std::string> GetNextFilename() {
-    auto basename = ::arrow::internal::Replace(write_options_.basename_template,
-                                               kIntegerToken, ToChars(file_counter_++));
+    std::optional<std::string> basename;
+    if (write_options_.basename_template_functor == nullptr) {
+      basename = ::arrow::internal::Replace(write_options_.basename_template,
+                                            kIntegerToken, ToChars(file_counter_++));
+    } else {
+      basename = ::arrow::internal::Replace(
```

Comment:
If a user is providing a `basename_template_functor` do we need to use `basename_template` at all?  Can this just be...

```
basename = write_options_.basename_template_functor(file_counter_++);
```

I don't think it matters too much but I think it would be simpler to say "either you use the template OR you use a custom function".

Reply:
good point, updated

Reply:
Ok.  I am happy keeping it how it is.  If a user wants to avoid basename_template they can just set it to `"{i}"`.